Add groups to citar-open functions

localauthor opened the pull request:
Changes:
- Adds `completing-read` to `citar--open-note`, for when multiple notes exist

- Changes `citar-select-files` to `citar-select-resources`, and calls the new function in both `citar-open` and `citar-open-library-files`

- Adds group-function, `citar-select-group-related-resources`, to `citar-select-resources`, to group candidates by type

The result of calling `citar-open` on an entry with associated files, notes, and a link (the following shows `vertico` with the `consult-completing-read-multiple` interface):
![citar-open file-link-note](https://user-images.githubusercontent.com/78040295/146552364-853bd6c3-6e5b-4bdd-aace-72dd418c2a96.jpg)

Selecting multiple file types, a note and a file:
![citar-open file-link-note selected](https://user-images.githubusercontent.com/78040295/146552354-8e7c0ce0-ff02-4597-918c-bd92c39cec78.jpg)


Todo:
~- Somehow set proper category metadata for links in `citar-select-resources`. I'm actually not sure if this is possible.
    - Note that prior to this PR, `citar-open` did not add any metadata to candidates, so `embark-act`'ing on a file candidate did not make embark file functions available. Now embark file functions are available, but on both files _and_ links. Neither case seems optimal, if it's not possible (or easy) to set category metadata by group.~


Question:
~- Does this work as expected with `embark-act-all`? I've not used that enough to properly test it.~

bdarcus commented on citar.el:
```diff
@@ -413,26 +413,36 @@ documentation for the return value and the meaning of
+  "Select resource(s) from a list of FILES, and optionally LINKS."
+  ;; DONE Add links to candidates
+  ;; DONE Add group-function
+  ;; FIX: set category metadata depending on type (differentiate links and files); possible?
```
Whenever I've contemplated something like this (for example with `citar-open`, this has been the hangup.

It is possible and easy to use `consult--multi`, which will do this. But then it's `completing-read` based, so only a single selected candidate.

I simply never figured out how to do it otherwise.

Maybe you can include a screenshot or two on the main post when you get a chance?

localauthor replied:
> It is possible and easy to use `consult--multi`, which will do this. But then it's `completing-read` based, so only a single selected candidate. And it's not explicit dependency.

This is where I landed. Bit of an impasse.

bdarcus commented:
@localauthor - I set this to merge to the `no-crm` branch, but before doing that, I'll put comments here:

1. I think the "Files" group maybe should be "LIbrary files"? WDYT?
2. I realize this when I was playing with this idea earlier: links have no annotations. Should they? If yes, what should they be?
3. What should happen to the existing more specific "open" commands? I guess we still need `citar-open-notes` at least?

Let me know your thoughts, and if you want to make any changes here before I merge?

bdarcus commented:
> Does this work as expected with `embark-act-all`? I've not used that enough to properly test it.

I am just starting to test this branch. Here's what you see after running `embark-act-all`:

![image](https://user-images.githubusercontent.com/1134/146783330-46203fa7-a21d-48d8-aa31-43540755b8ab.png)

Two things:

1. I guess notes should also have a `file` category, but remain a separate group? If yes, not sure how to do that.
2. I'm actually not sure how one would then open all of these results from there using `embark-act-all`.

Any thoughts on the above @minad?

localauthor commented:
Ok, I just rebased onto main. 

I changed `citar--open-notes` to use `citar-select-resources`, so now notes are properly categorized as files and grouped as notes.

As for the question of where to merge this (either to `main` directly or to `no-crm`) , I suppose it depends on your plans for `no-crm`. :)

This branch changes CRM to CR in all the open functions anyway.

bdarcus commented:
Let's keep it here for now. There are still some subtleties with `embark-act-all` to work out.

bdarcus commented:
> Per my point 2 [above](https://github.com/bdarcus/citar/pull/491#issuecomment-997980185), I can't see how to run `embark-act-all` to open all of the resulting resources, because the default action is then `embark-act`. I wonder if that needs to be the case, or what a better alternative might be.

There does seem to be a bug here though. If I only select one candidate, and then do `embark-act-all` on this results, where the default action is `citar-open`, I get a "no results found".

Even if I only select one resource, I get a "wrong type" error (it's expecting a list).

bdarcus commented on citar.el:
```diff
@@ -885,7 +896,7 @@ into the corresponding reference key.  Return
 (defun citar-open (keys-entries)
-  "Open related resources (links or files) for KEYS-ENTRIES."
+  "Open any related resources (links, files, or notes) for KEYS-ENTRIES."
   (interactive (list (citar-select-refs
```
isn't this still CRM?

localauthor replied:
Only if `citar-select-multiple` is set to `t`, I think?

bdarcus commented on citar.el:
```diff
@@ -419,26 +419,37 @@ documentation for the return value and the meaning of
-    (let ((extension (file-name-extension file)))
-      (when transform file
-        ;; Transform for grouping and group title display.
+(defun citar-select-resources (files &optional links)
```
if this is now CR, shouldn't this be singular?
```suggestion
(defun citar-select-resource (files &optional links)
```

bdarcus commented on citar.el:
```diff
@@ -885,7 +896,7 @@ into the corresponding reference key.  Return
 ;;;###autoload
 (defun citar-open (keys-entries)
-  "Open related resources (links or files) for KEYS-ENTRIES."
+  "Open any related resources (links, files, or notes) for KEYS-ENTRIES."
```
```suggestion
  "Open any related resource (links, files, or notes) for KEYS-ENTRIES."
```

localauthor commented:
All the `open` functions should be working as expected now (hopefully!), with the exception of `embark-act-all`, which I'm just starting to work through.

bdarcus commented:
> All the `open` functions should be working as expected now (hopefully!), with the exception of `embark-act-all`, which I'm just starting to work through.

When I use `citar-open`, I still get a type error when trying to open any resource.

localauthor commented:
Latest commit make `embark-act-all` work appropriately in all contexts.

As the commit message says, each command now let-binds `embark-default-action-overrides` to change the default action from the original commands (ie, `citar-open`, `citar-open-notes`, etc.) which cannot act on files or urls, to functions that _can_ act on files and urls, namely, `citar-open-file`, `find-file` and `browse-url`, where appropriate.

The edge case here is with `citar-open` and acting on candidates of multiple types (files or urls). I've added a function `citar-open-multi` to deal with candidates that are of the type `consult-multi`. So now running `embark-act-all` on candidates that include links and files should open all of them as expected.

Hopefully it works for you as it does for me.

localauthor commented:
One issue with using `consult-multi` is that it requires `embark-consult.el` to be loaded and, specifically, `embark-consult--multi-transform` to be defined.

If only embark.el is loaded, these are not available.

bdarcus commented:
> One issue with using `consult-multi` is that it requires `embark-consult.el` to be loaded and, specifically, `embark-consult--multi-transform` to be defined.
> 
> If only `embark.el` is loaded, these are not available.

What if you define the variable? Do you get an error, or just not the behavior?

localauthor commented:
`embark-consult--multi-transform` is a function, so it really needs to loaded for everything to fully work.

I say "fully" because `citar-open` still functions reasonably without it, ie, there are no errors and files/links can be opened with "RET", according to type --- even when using `embark-act-all` on multiple types.

Since without `embark-consult--multi-transform`, files and urls remain typed as 'consult-multi', they are handled by the default action associated with that type, namley, the new 'citar-open-multi'. There aren't file- or url-specific actions available, but each is opened appropriately, and there are no errors.

bdarcus commented:
> Each `open` command now includes a let-binding for `embark-default-action-overrides` in order to override the default action depending on target type.

These result in "unused lexical variable" warnings.

On a separate matter, it's getting closer. But if I have this:

![image](https://user-images.githubusercontent.com/1134/146937970-b0dc96b9-074b-49c4-9cfd-53db4eb06499.png)

... and then do `embark-act-all`, I get this:

![image](https://user-images.githubusercontent.com/1134/146938073-d79e1c44-defb-462b-8df4-7ca26df6aad1.png)

minad commented:
@localauthor Yes of course. We should switch to `multi` from `consult-multi` to make this general and to avoid the loading issue. But I will wait for a few weeks to let this settle here, since Citar does not yet seem to have converged fully. Then we can change it across the different packages and tag new releases. Probably embark and marginalia should support both `multi` and `consult-multi` for some time and treat them as aliases.

localauthor commented:
> ... and then do `embark-act-all`, I get this:

Before pressing "RET", what does the embark-indicator-menu look like for you? That is: what is the default action you are calling when do `embark-act-all`? If it is `citar-open`, then I would expect to see exactly what you're screenshot shows (which is a symptom of another issue I'm trying to figure out, about which more soon).

Here is what I see _before_ pressing "RET" in a similar series of steps:
![Screen Shot 2021-12-21 at 15 33 20](https://user-images.githubusercontent.com/78040295/146938930-72454c95-ea03-422d-8abb-6e60c5d933ee.jpg)

localauthor commented:
> These result in "unused lexical variable" warnings.

I'm afraid I don't know how to avoid such warnings, or what they mean really.

bdarcus commented:
> Followed by opened files, notes, and browser.

I see the same, except for this last step.

Maybe there's something wrong with my install of the branch. Running `eval-buffer` on `citar.el` does result in the correct behavior!

bdarcus commented:
> > These result in "unused lexical variable" warnings.
> 
> I'm afraid I don't know how to avoid such warnings, or what they mean really.

@minad - any tips? He's let-binding `embark-default-action-overrides` per https://github.com/bdarcus/citar/pull/491#issuecomment-998660728.

minad commented:
Oh. These default action overrides should not be necessary. The default file and url actions should work ootb as should all the other actions. If not something else is wrong and should be fixed.

The correct approach is to define your own category if you want to overwrite or define your own actions.

bdarcus commented:
> Probably you miss a defvar to fix the lexical error?

Yes; I just committed that change.

minad commented on citar.el:
```diff
@@ -903,17 +918,21 @@ into the corresponding reference key.  Return
+              (t (citar-file-open selection)))
       (message "No associated resources"))))
 
+(defun citar-open-multi (selection)
```
This command is unnecessary and should be removed. Embark already dispatches consult-multi to file and url. That's the purpose of consult-multi.

minad commented on citar.el:
```diff
@@ -885,13 +897,16 @@ into the corresponding reference key.  Return
-  (let* ((key-entry-alist (citar--ensure-entries keys-entries))
+  (let* ((embark-default-action-overrides '((consult-multi . citar-open-multi)
+					    (file . citar-file-open)
+                                            (url . browse-url)))
```
Here the consult-multi and url overrides should probably go away.

localauthor replied:
Thanks for looking at this, and for the comments.

Embark and consult-multi work just as you describe when it comes to categorizing files and urls, but (as I describe [here](https://github.com/bdarcus/citar/pull/491#issuecomment-998660728)) because the source of the target is minibuffer completion, Embark sets the default action for all targets, regardless of type, to the command that opened the minibuffer in the first place (per the docstring of `embark-default-action-overrides`). That means the default action for any target type is set to `citar open`, which is a function that does not take files or urls as arguments. Hence the overrides. It's a bit of a hail mary, I guess, but I'm afraid I'm running up against the limits of my lisp/emacs knowledge here. ;)

As for the `citar-open-multi` function, it was only added to allow `embark-act-all` to work on properly on targets of multiple types. When using `embark-act` on single candidates, consult-multi works just as expected and this function isn't called. When calling `embark-act-all` on targets of multiple types, they haven't been transformed by the candidate transformer yet ---  so they remain typed as `consult-multi`. Hence, another override. Again, bit of a hail mary -- but functional!

Thanks again for your comments. Really appreciate your taking the time.

minad replied:
Okay, maybe I missed something here but this seems wrong. The default action is supposed to be *the command that started the completion*. Overriding it with something else is counterintuitive and goes against the usual Embark behavior.

Note that Embark has a minibuffer argument injection mechanism. I suggest you dig deeper into the Embark documentation.

minad replied:
> That means that, in this case, the default action for any target type is set to citar open, which is a function that does not take files or urls as arguments. Hence the overrides. 

I am specifically referring to this. This is wrong. There is the minibuffer injection mechanism.

> When calling embark-act-all on targets of multiple types, they haven't been transformed by the candidate transformer yet --- so they remain typed as consult-multi.

I will check this, but if this is the case it is a bug or issue in Embark. 

I know that oantolin added some target type checking to embark-act-all and there is indeed a problem when acting on candidates of multiple categories at once. Anyways this should be handled properly by  Embark itself.

bdarcus replied:
I experimented a bit with removing these lines, and things seem to work fine, until I run `embark-act-all` on the resulting resources. So to recap:

1. `citar-open`
2. narrow candidates and then `embark-act-all`
3. from there I get a list of resources, if I run `embark-act-all`, II then get the error below:

```elisp
embark PCH: (user-error "No target found")
```

... which is I think why he introduced the `citar-open-multi` function.

![image](https://user-images.githubusercontent.com/1134/147260488-87902c82-b8aa-48aa-a75c-c1065c8bc725.png)

minad replied:
Did you load embark-consult?

bdarcus replied:
> Did you load embark-consult? 

I did require it now, and now get:

```elisp
(execute-extended-command nil "citar-open" nil)
```

> I don't know what is wrong here exactly, but the override should not be needed due to the design of embark. consult-buffer also works without override.

I'll take another look.

minad replied:
@bdarcus What's the status here? Did you figure out the issues?

bdarcus replied:
> Did you figure out the issues?

Not yet. Likely in the next few days.

minad commented:
@bdarcus As a small demonstrator, this works for me:

~~~ elisp
(require 'embark-consult)
(defun open-example ()
  (interactive)
  (let* ((candidates
          (list (propertize "init.el" 'consult-multi
                            '(file . "~/.config/emacs/init.el"))
                (propertize "emacs.org" 'consult-multi
                            '(url . "https://www.gnu.org/software/emacs/"))))
         (result (completing-read "Open: "
                                  (lambda (str pred action)
                                    (if (eq action 'metadata)
                                        '(metadata (category . consult-multi))
                                      (complete-with-action action candidates str pred)))
                                  nil t)))
    (pcase (get-text-property 0 'consult-multi (car (member result candidates)))
      (`(url . ,url) (eww url))
      (`(file . ,file) (find-file file)))))
~~~
